c++/test/TestReader.cc

```diff
@@ -250,4 +252,206 @@ namespace orc {
+    verifyUnionSelection({{2, ReadIntent_OFFSETS}}, {0, 2});
+
+    // select only the offsets of single_union and the outermost offsets of nested_union.
+    verifyUnionSelection({{2, ReadIntent_OFFSETS}, {5, ReadIntent_OFFSETS}}, {0, 2, 5});
+
+    // select only the offsets of single_union and the innermost offsets of nested_union.
+    verifyUnionSelection({{2, ReadIntent_OFFSETS}, {8, ReadIntent_OFFSETS}},
+                         {0, 2, 5, 6, 7, 8, 11});
+  }
+
+  TEST(TestReadIntent, testUnionAllAndOffsets) {
+    // select all of single_union and only the outermost offsets of nested_union.
+    verifyUnionSelection({{2, ReadIntent_ALL}, {5, ReadIntent_OFFSETS}}, {0, 2, 3, 4, 5});
+  }
+
+  TEST(TestReadIntent, testUnionConflictingIntent) {
+    // test conflicting ReadIntent on nested_union.
+    verifyUnionSelection({{5, ReadIntent_OFFSETS}, {8, ReadIntent_ALL}},
+                         {0, 5, 6, 7, 8, 9, 10, 11});
+    verifyUnionSelection({{5, ReadIntent_ALL}, {8, ReadIntent_OFFSETS}},
+                         {0, 5, 6, 7, 8, 9, 10, 11});
+    verifyUnionSelection({{5, ReadIntent_OFFSETS}, {6, ReadIntent_ALL}, {8, ReadIntent_OFFSETS}},
+                         {0, 5, 6, 7, 8, 9, 10, 11});
+  }
+
+  TEST(TestReadIntent, testUnionRowBatchContent) {
+    std::unique_ptr<Reader> reader = createExampleReader("TestOrcFile.nestedUnion.orc");
+
+    // select all of single_union and only the offsets of nested_union.
+    RowReaderOptions::IdReadIntentMap idReadIntentMap = {{2, ReadIntent_ALL},
+                                                         {5, ReadIntent_OFFSETS}};
+    RowReaderOptions rowReaderOpts;
+    rowReaderOpts.includeTypesWithIntents(idReadIntentMap);
+    std::unique_ptr<RowReader> rowReader = reader->createRowReader(rowReaderOpts);
+
+    // Read a row batch.
+    std::unique_ptr<ColumnVectorBatch> batch = rowReader->createRowBatch(1024);
+    EXPECT_TRUE(rowReader->next(*batch));
+    EXPECT_EQ(4, batch->numElements);
+    auto& structBatch = dynamic_cast<StructVectorBatch&>(*batch);
+
+    // verify content of int_array selection.
+    auto& unionBatch = dynamic_cast<UnionVectorBatch&>(*structBatch.fields[0]);
+    EXPECT_EQ(4, unionBatch.numElements);
+    EXPECT_EQ(2, unionBatch.children.size());
+    auto& longBatch = dynamic_cast<LongVectorBatch&>(*unionBatch.children[0]);
+    auto& stringBatch = dynamic_cast<StringVectorBatch&>(*unionBatch.children[1]);
+    EXPECT_EQ(3, longBatch.numElements);
+    EXPECT_EQ(1, stringBatch.numElements);
+    // verify content of the first row.
+    EXPECT_EQ(0, unionBatch.tags.data()[0]);
+    EXPECT_EQ(0, unionBatch.offsets.data()[0]);
+    EXPECT_EQ(0, longBatch.data.data()[0]);
+    // verify content of the second row.
+    EXPECT_EQ(1, unionBatch.tags.data()[1]);
+    EXPECT_EQ(0, unionBatch.offsets.data()[1]);
+    EXPECT_EQ(2, stringBatch.length.data()[0]);
+    EXPECT_EQ(0, strncmp("s1", stringBatch.data.data()[0], 2));
```

Comment:
I feel like it's more readable if we create the test ORC files in flight. E.g. https://github.com/apache/orc/blob/0622002bd140d7bd88af0ae71d73fd1b3a101c17/c%2B%2B/test/TestPredicatePushdown.cc#L29

Otherwise, each developer should manually dump the file content to understand these. So I think we should try creating the ORC files in tests unless we can't (e.g. for corrupt files). Creating the ORC files inflight also reduce the repo size. But I'm not strongly opposing the current approach.
@dongjoon-hyun Could you share your opinion?

Reply:
+1 for @stiga-huang 's comment about creating the test file in the test cases.

Thank you for your review, @stiga-huang .

Reply:
Done